[CWS] add support for socket system call

loresuso opened the pull request:
### What does this PR do?

This PRs attempts to add support to the `socket` system call

### Motivation

There is the possibility to write rules and expressions on this system event, like for instance opening a `SOCK_RAW` socket.

loresuso committed:
new: monitor socket syscall in eBPF

loresuso committed:
feat: add socket event model and serialization

loresuso committed:
feat: add socket event support for eBPFless mode

loresuso committed:
test: socket event

loresuso committed:
feat: add socket event approvers and kfilters

loresuso committed:
chore: update generated code for socket event

loresuso committed:
chore: add release note for socket syscall support

loresuso committed:
fix: remove wrong hook_sys_socket selectors

NetworkSelectors referenced hookFunc("hook_sys_socket") and
hookFunc("rethook_sys_socket") — function names that do not exist in
the compiled eBPF object. HOOK_SYSCALL_ENTRY3(socket) generates
kprobe__{32,64}_sys_socket and kretprobe__{32,64}_sys_socket sections,
not hook_sys_socket. The socket syscall probes are correctly selected
elsewhere via ExpandSyscallProbesSelector under the "socket" event
type, so these entries were redundant as well as wrong. They caused
KMT TestEventMonitor / TestActivityDumps / TestSecurityProfile* / the
policy reload path in TestRemediationCustomEvents to fail with:
  couldn't find program {UID:security EBPFFuncName:hook_sys_socket}

loresuso commented:
@codex review

theop-dd commented on pkg/security/ebpf/c/include/hooks/network/socket.h:
```diff
@@ -1,6 +1,73 @@
+    // Mask out SOCK_CLOEXEC / SOCK_NONBLOCK flags to get just the socket type
+    // (SOCK_STREAM=1, SOCK_DGRAM=2, ... fit in 4 bits)
+    u16 socket_type = (u16)type & 0x0F;
+    return trace__sys_socket(SYNC_SYSCALL, (u16)domain, socket_type, (u16)protocol);
```
Can't we have an issue / lose info if we truncate the domain (which is an int)?

loresuso replied:
Linux defines only ~40 socket domains (`AF_MAX` is around 46) and a similarly small number of types/protocols, so the values fit comfortably in `u16` without any loss of information.

theop-dd commented on pkg/security/ebpf/c/include/maps.h:
```diff
@@ -37,6 +37,9 @@ BPF_ARRAY_MAP(sysctl_action_approvers, struct u32_flags_filter_t, 1)
 BPF_ARRAY_MAP(setsockopt_level_or_optname_approvers, struct u64_flags_filter_t, 2)
+BPF_ARRAY_MAP(socket_domain_approvers, struct u64_flags_filter_t, 1)
+BPF_ARRAY_MAP(socket_protocol_approvers, struct u64_flags_filter_t, 1)
+BPF_ARRAY_MAP(socket_type_approvers, struct u64_flags_filter_t, 1)
```
nit: This should work perfectly like that but maybe we could do something similar we did for setsockopt and put the 3 approvers in the same map to avoid multiplying the maps since they all are flags approvers. WDYT?

theop-dd commented on pkg/security/probe/monitors/approver/approver_monitor.go:
```diff
@@ -126,7 +126,7 @@ func (d *Monitor) SendStats() error {
 		}
 	}
-	for i := uint32(0); i != uint32(model.LastApproverEventType); i++ {
+	for i := uint32(0); i <= uint32(model.LastApproverEventType); i++ {
```
I would recommend to open another PR for this fix so we can keep track of it.

theop-dd commented on pkg/security/ptracer/network_handlers.go:
```diff
@@ -191,21 +190,37 @@ func handleConnect(tracer *Tracer, process *Process, msg *ebpfless.SyscallMsg, r
+	domain := uint16(tracer.ReadArgInt32(regs, 0))
+	rawSockType := tracer.ReadArgInt32(regs, 1)
+	// Mask out SOCK_CLOEXEC / SOCK_NONBLOCK flags; SOCK_STREAM=1, SOCK_DGRAM=2, ... fit in 4 bits.
+	sockType := uint16(rawSockType) & 0x0F
```
nit: not sure the & op is useful since we already have a uint16

loresuso replied:
The `& 0x0F` mask isn't redundant with the `uint16` cast: `SOCK_NONBLOCK` is `0x800` and `SOCK_CLOEXEC` is `0x80000`, and `SOCK_NONBLOCK` (0x800) fits inside a `uint16` and would survive the cast. The mask is what strips those flags to leave just the socket type (SOCK_STREAM=1, SOCK_DGRAM=2, ... fit in 4 bits). Same masking is done on the eBPF side in `socket.h`.

theop-dd replied:
Oh yes now I get it, thanks for the explanation!

theop-dd commented on pkg/security/secl/model/model_unix.go:
```diff
@@ -125,6 +125,7 @@ type Event struct {
 	Connect    ConnectEvent    `field:"connect" event:"connect"`       // [7.60] [Network] A connect was executed
 	Accept     AcceptEvent     `field:"accept" event:"accept"`         // [7.63] [Network] An accept was executed
 	SetSockOpt SetSockOptEvent `field:"setsockopt" event:"setsockopt"` // [7.68] [Network] A setsockopt was executed
+	Socket     SocketEvent     `field:"socket" event:"socket"`         // [7.68] [Network] A socket was created
```
You can change the version to 7.80

theop-dd commented on pkg/security/probe/kfilters/socket.go:
```diff
@@ -0,0 +1,64 @@
+	},
+	{
+		Field:        "socket.type",
+		TypeBitmask:  eval.ScalarValueType | eval.BitmaskValueType,
```
I'm not sure about the BtmaskValueType here. Do you want to use the syntaxe socket.type & type > 0 ?

loresuso replied:
you are right!

theop-dd commented on pkg/security/ptracer/network_handlers.go:
```diff
@@ -191,21 +190,37 @@ func handleConnect(tracer *Tracer, process *Process, msg *ebpfless.SyscallMsg, r
 		case unix.SOCK_DGRAM:
-			socketMsg.Protocol = unix.IPPROTO_UDP
-		default:
+			internalProtocol = unix.IPPROTO_UDP
```
I'm not a big fan of putting UDP as the default value. I think keeping 0 would be easier to identify if there is an issue in the future. WDYT?

loresuso replied:
I think 0 is the default for TCP though

loresuso replied:
I think here we should infer 0 -> to TCP. In kernel we would read IPPROTO_TCP, and should be done also here probably, but still having a look. 
socket(AF_INET, SOCK_STREAM, 0) == socket(AF_INET, SOCK_STREAM, IPPROTO_TCP), but the value for IPPROTO_TCP is 6. So there should be something that does this

theop-dd replied:
0 is not a specific value. It means that the kernel will try to guess the protocol. If he has SOCK_STREAM he will guess TCP, if he has SOCK_DGRAM he will guess UDP. 0 is not related to any specific protocol. THat's what we are trying to compensate here with this if else afaik.

theop-dd commented on pkg/security/secl/model/model_unix.go:
```diff
@@ -125,6 +125,7 @@ type Event struct {
 	Connect    ConnectEvent    `field:"connect" event:"connect"`       // [7.60] [Network] A connect was executed
 	Accept     AcceptEvent     `field:"accept" event:"accept"`         // [7.63] [Network] An accept was executed
 	SetSockOpt SetSockOptEvent `field:"setsockopt" event:"setsockopt"` // [7.68] [Network] A setsockopt was executed
+	Socket     SocketEvent     `field:"socket" event:"socket"`         // [7.80] [Network] A socket was created
```
It will be 7.81, sorry xd ^^

loresuso committed:
refactor: address review feedback on socket event

- Drop BitmaskValueType from the socket.type capability: socket types
  are scalar enum values (SOCK_STREAM, SOCK_DGRAM, ...) and the
  SOCK_NONBLOCK / SOCK_CLOEXEC flags are already masked out in the
  eBPF hook, so bitmask matching is not meaningful here.
- Bump the from_agent_version annotation for socket from 7.80 to 7.81
  to match the upcoming release.

theop-dd committed:
add static

loresuso commented:
/merge -m squash
